python/ray/tune/execution/tune_controller.py

```diff
@@ -0,0 +1,1196 @@
+            self._remove_actor(tracked_actor=tracked_actor)
+
+    def _cleanup_stopping_actors(self, force_all: bool = False):
+        now = time.monotonic()
+
+        if (
+            not force_all
+            and now - self._earliest_stopping_actor > self._actor_cleanup_timeout
+        ):
+            # If the earliest actor to timeout has not reached the timeout, return
+            return
+
+        # This is a bit costly, so we want to avoid running it too often
+        times = deque(
+            sorted(
+                (timestamp, tracked_actor)
+                for tracked_actor, timestamp in self._stopping_actors.items()
+            )
+        )
+
+        while times and (
+            force_all or time.monotonic() - times[0][0] > self._actor_cleanup_timeout
+        ):
+            _, tracked_actor = times.popleft()
+            if self._actor_manager.is_actor_started(tracked_actor=tracked_actor):
+                self._actor_manager.remove_actor(tracked_actor=tracked_actor, kill=True)
+            self._stopping_actors.pop(tracked_actor)
+
+        if times:
+            self._earliest_stopping_actor = times[0][0]
+        else:
+            self._earliest_stopping_actor = float("inf")
+
+    def step(self):
+        if self.is_finished():
+            raise TuneError("Called step when all trials finished?")
+
+        with warn_if_slow("on_step_begin"):
+            self.on_step_begin()
+
+        with warn_if_slow("callbacks.on_step_begin"):
+            self._callbacks.on_step_begin(
+                iteration=self._iteration, trials=self._trials
+            )
+
+        # Ask searcher for more trials
+        self._maybe_update_trial_queue()
+
+        # Start actors for added trials
+        self._maybe_add_actors()
+
+        # Handle one event
+        if not self._actor_manager.next(timeout=0.1):
+            # If there are no actors running, warn about potentially
+            # insufficient resources
+            if not self._actor_manager.num_live_actors:
+                self._insufficient_resources_manager.on_no_available_trials(
+                    self.get_trials()
+                )
+
+        # Maybe stop whole experiment
+        self._stop_experiment_if_needed()
+
+        # Maybe save experiment state
+        try:
+            self.checkpoint()
+        except Exception as e:
+            logger.warning(f"Trial controller checkpointing failed: {str(e)}")
+            raise e
+
+        self._iteration += 1
+
+        if self._server:
+            with warn_if_slow("server"):
+                self._process_stop_requests()
+
+            if self.is_finished():
+                self._server.shutdown()
+
+        with warn_if_slow("on_step_end"):
+            self.on_step_end()
+        with warn_if_slow("callbacks.on_step_end"):
+            self._callbacks.on_step_end(iteration=self._iteration, trials=self._trials)
+
+    def _set_trial_status(self, trial: Trial, status: str):
```

Comment:
Is this the canonical way for schedulers to interact with trial status (not to say that they should... but some of them are) in the future?

Reply:
Yes! The TuneController should be the sole interface for modifying any training/tuning behavior. I think we should prevent setting trial status manually in the future (separate from this PR)